prow/cmd/exporter/main.go

```diff
@@ -0,0 +1,128 @@
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	configPath     string
+	jobConfigPath  string
+	selectorString string
+
+	kubernetes prowflagutil.KubernetesOptions
+
+	selector labels.Selector
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+
+	fs.StringVar(&o.configPath, "config-path", "", "Path to config.yaml.")
+	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
+	fs.StringVar(&o.selectorString, "label-selector", labels.Everything().String(), "Label selector to be applied in prowjobs. See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors for constructing a label selector.")
+
+	for _, group := range []flagutil.OptionGroup{&o.kubernetes} {
+		group.AddFlags(fs)
+	}
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatalf("cannot parse args: '%s'", os.Args[1:])
+	}
+	o.configPath = config.ConfigPath(o.configPath)
+	return o
+}
+
+func (o *options) Validate() error {
+	for _, group := range []flagutil.OptionGroup{&o.kubernetes} {
+		if err := group.Validate(false); err != nil {
+			return err
+		}
+	}
+
+	selector, err := labels.Parse(o.selectorString)
+	if err != nil {
+		return fmt.Errorf("parse label selector: %v", err)
+	}
+	o.selector = selector
+	return nil
+}
+
+func mustRegister(component string, selector labels.Selector, lister lister) *prometheus.Registry {
+	registry := prometheus.NewRegistry()
+	prometheus.WrapRegistererWith(prometheus.Labels{"collector_name": component}, registry).MustRegister(&prowJobCollector{
+		name:     component,
+		lister:   lister,
+		selector: selector,
+	})
+	registry.MustRegister(
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+		prometheus.NewGoCollector(),
+	)
+	return registry
+}
+
+func main() {
+	logrusutil.ComponentInit("exporter")
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	defer interrupts.WaitForGracefulShutdown()
+
+	pjutil.ServePProf()
+
+	configAgent := &config.Agent{}
+	if err := configAgent.Start(o.configPath, o.jobConfigPath); err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+	cfg := configAgent.Config
+
+	pjClientset, err := o.kubernetes.ProwJobClientset(cfg().ProwJobNamespace, false)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create prowjob client set")
+	}
+	informerFactory := prowjobinformer.NewSharedInformerFactory(pjClientset, 15*time.Second)
+	pjLister := informerFactory.Prow().V1().ProwJobs().Lister()
+	go informerFactory.Start(interrupts.Context().Done())
+
+	registry := mustRegister("exporter", o.selector, pjLister)
+
+	// Expose prometheus metrics
+	metrics.ExposeMetricsWithRegistry("exporter", cfg().PushGateway, registry)
+
+	logrus.Info("exporter is running ...")
+	select {} // block forever
```

Comment:
You need `interrupts.WaitForGratefulShutdown`

Reply:
Good catch.
I have that one already but have not realized its function.
Removed `select`